tests/csapi/apidoc_room_members_test.go

```diff
@@ -107,5 +108,123 @@ func TestRoomMembers(t *testing.T) {
+				JSON: []match.JSON{
+					match.JSONKeyPresent("room_id"),
+					match.JSONKeyTypeEqual("room_id", gjson.String),
+					match.JSONKeyEqual("room_id", roomID),
+				},
+			})
+
+			bob.SyncUntilTimelineHas(
+				t,
+				roomID,
+				func(ev gjson.Result) bool {
+					if ev.Get("type").Str != "m.room.member" || ev.Get("state_key").Str != bob.UserID {
+						return false
+					}
+					must.EqualStr(t, ev.Get("content").Get("membership").Str, "join", "Bob failed to join the room")
+					must.EqualStr(t, ev.Get("content").Get("foo").Str, "bar", "Failed to propagate custom content")
+					return true
+				},
+			)
+		})
+		// sytest: POST /join/:room_alias can join a room with custom content
+		t.Run("POST /join/:room_alias can join a room with custom content", func(t *testing.T) {
+			t.Parallel()
+			roomID := alice.CreateRoom(t, map[string]interface{}{
+				"visibility":      "public",
+				"preset":          "public_chat",
+				"room_alias_name": "room_alias_random_2",
+			})
+
+			reqBody := client.WithJSONBody(t, map[string]interface{}{
+				"foo": "bar",
+			})
+
+			res := bob.MustDoFunc(t, "POST", []string{"_matrix", "client", "r0", "join", "#room_alias_random_2:hs1"}, reqBody)
+
+			must.MatchResponse(t, res, match.HTTPResponse{
+				JSON: []match.JSON{
+					match.JSONKeyPresent("room_id"),
+					match.JSONKeyTypeEqual("room_id", gjson.String),
+					match.JSONKeyEqual("room_id", roomID),
+				},
+			})
+
+			bob.SyncUntilTimelineHas(
+				t,
+				roomID,
+				func(ev gjson.Result) bool {
+					if ev.Get("type").Str != "m.room.member" || ev.Get("state_key").Str != bob.UserID {
+						return false
+					}
+					must.EqualStr(t, ev.Get("content").Get("membership").Str, "join", "Bob failed to join the room")
+					must.EqualStr(t, ev.Get("content").Get("foo").Str, "bar", "Failed to propagate custom content")
+					return true
+				},
+			)
+		})
+		// sytest: POST /rooms/:room_id/leave can leave a room
+		t.Run("POST /rooms/:room_id/leave can leave a room", func(t *testing.T) {
+			t.Parallel()
+			roomID := alice.CreateRoom(t, map[string]interface{}{
+				"visibility": "public",
+				"preset":     "public_chat",
+			})
+
+			res := bob.MustDoFunc(t, "POST", []string{"_matrix", "client", "r0", "join", roomID})
+
+			must.MatchResponse(t, res, match.HTTPResponse{
+				JSON: []match.JSON{
+					match.JSONKeyPresent("room_id"),
+					match.JSONKeyTypeEqual("room_id", gjson.String),
+					match.JSONKeyEqual("room_id", roomID),
+				},
+			})
+
+			bob.SyncUntilTimelineHas(
+				t,
+				roomID,
+				func(ev gjson.Result) bool {
+					if ev.Get("type").Str != "m.room.member" || ev.Get("state_key").Str != bob.UserID {
+						return false
+					}
+					must.EqualStr(t, ev.Get("content").Get("membership").Str, "join", "Bob failed to join the room")
+					return true
+				},
+			)
+
+			_ = bob.MustDoFunc(t, "POST", []string{"_matrix", "client", "r0", "rooms", roomID, "leave"})
+
+			bob.SyncUntilTimelineHas(
```

Comment:
@kegsay I am getting a failure here with logs
```sh
=== CONT  TestRoomMembers
    client.go:370: GET hs1/_matrix/client/r0/sync => 200 OK (1.016855872s)
=== CONT  TestRoomMembers/Parallel/POST_/rooms/:room_id/leave_can_leave_a_room
    apidoc_room_members_test.go:214: HTTP/1.1 200 OK
        Transfer-Encoding: chunked
        Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, Date
        Access-Control-Allow-Methods: GET, HEAD, POST, PUT, DELETE, OPTIONS
        Access-Control-Allow-Origin: *
        Cache-Control: no-cache, no-store, must-revalidate
        Content-Type: application/json
        Date: Tue, 17 Aug 2021 19:01:51 GMT
        Server: Synapse/1.37.1

        157
        {"account_data":{"events":[]},"to_device":{"events":[]},"device_lists":{"changed":[],"left":[]},"presence":{"events":[]},"rooms":{"join":{},"invite":{},"knock":{},"leave":{}},"groups":{"join":{},"invite":{},"leave":{}},"device_one_time_keys_count":{},"org.matrix.msc2732.device_unused_fallback_key_types":[],"next_batch":"s22_3_0_1_1_1_1_3_1"}
        0

    apidoc_room_members_test.go:214: SyncUntil: timed out. Called check function 0 times
2021/08/18 00:31:52 ============================================


2021/08/18 00:31:52 227d4b5cbe443c7cd67db01bcb6bc6db3c02b78a6f640d9a4f69bc84da188624 : Server logs:
Signature ok
subject=CN = hs1
Getting CA Private Key
Config is missing macaroon_secret_key
2021-08-17 19:01:45,348 - root - 323 - WARNING - main - ***** STARTING SERVER *****
2021-08-17 19:01:45,348 - root - 324 - WARNING - main - Server /usr/local/lib/python3.8/site-packages/synapse/app/homeserver.py version 1.37.1
2021-08-17 19:01:45,348 - root - 325 - INFO - main - Server hostname: hs1
2021-08-17 19:01:45,349 - root - 326 - INFO - main - Instance name: master
2021-08-17 19:01:45,349 - synapse.app.homeserver - 360 - INFO - main - Setting up server
2021-08-17 19:01:45,349 - synapse.server - 309 - INFO - main - Setting up.
2021-08-17 19:01:45,359 - synapse.storage.databases - 49 - INFO - main - [database config 'master']: Checking database server
2021-08-17 19:01:45,360 - synapse.storage.databases - 52 - INFO - main - [database config 'master']: Preparing for databases ['main', 'state']
2021-08-17 19:01:45,361 - synapse.storage.prepare_database - 116 - INFO - main - ['main', 'state']: Checking existing schema version
2021-08-17 19:01:45,364 - synapse.storage.prepare_database - 120 - INFO - main - ['main', 'state']: Existing schema is 59 (+18 deltas)
2021-08-17 19:01:45,365 - synapse.storage.databases.main - 353 - INFO - main - Checking database for consistency with configuration...
2021-08-17 19:01:45,366 - synapse.storage.prepare_database - 407 - INFO - main - Applying schema deltas for v59
2021-08-17 19:01:45,367 - synapse.storage.prepare_database - 535 - INFO - main - Schema now up to date
2021-08-17 19:01:45,372 - synapse.storage.databases - 67 - INFO - main - [database config 'master']: Starting 'main' database
2021-08-17 19:01:45,372 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for public_room_list_stream(stream_id)
2021-08-17 19:01:45,373 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for device_lists_stream(stream_id)
2021-08-17 19:01:45,373 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for user_signature_stream(stream_id)
2021-08-17 19:01:45,374 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for device_lists_outbound_pokes(stream_id)
2021-08-17 19:01:45,374 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for e2e_cross_signing_keys(stream_id)
2021-08-17 19:01:45,374 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for event_reports(id)
2021-08-17 19:01:45,375 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for push_rules(id)
2021-08-17 19:01:45,375 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for push_rules_enable(id)
2021-08-17 19:01:45,376 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for local_group_updates(stream_id)
2021-08-17 19:01:45,376 - synapse.config.appservice - 56 - WARNING - main - Expected None to be a list of AS config files.
2021-08-17 19:01:45,377 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for receipts_linearized(stream_id)
2021-08-17 19:01:45,377 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for room_account_data(stream_id)
2021-08-17 19:01:45,377 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for room_tags_revisions(stream_id)
2021-08-17 19:01:45,378 - synapse.storage.databases.main.event_push_actions - 537 - INFO - main - Searching for stream ordering 1 month ago
2021-08-17 19:01:45,379 - synapse.storage.databases.main.event_push_actions - 541 - INFO - main - Found stream ordering 1 month ago: it's 2
2021-08-17 19:01:45,379 - synapse.storage.databases.main.event_push_actions - 544 - INFO - main - Searching for stream ordering 1 day ago
2021-08-17 19:01:45,380 - synapse.storage.databases.main.event_push_actions - 548 - INFO - main - Found stream ordering 1 day ago: it's 2
2021-08-17 19:01:45,381 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for events(stream_ordering)
2021-08-17 19:01:45,382 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for events(stream_ordering)
2021-08-17 19:01:45,382 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for ex_outlier_stream(event_stream_ordering)
2021-08-17 19:01:45,385 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for device_inbox(stream_id)
2021-08-17 19:01:45,390 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for pushers(id)
2021-08-17 19:01:45,390 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for deleted_pushers(stream_id)
2021-08-17 19:01:45,391 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for push_rules_stream(stream_id)
2021-08-17 19:01:45,396 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for presence_stream(stream_id)
2021-08-17 19:01:45,398 - synapse.storage.util.id_generators - 55 - INFO - main - initialising stream generator for access_tokens(id)
2021-08-17 19:01:45,399 - synapse.storage.databases - 84 - INFO - main - [database config 'master']: Starting 'state' database
2021-08-17 19:01:45,400 - synapse.storage.databases - 99 - INFO - main - [database config 'master']: prepared
2021-08-17 19:01:45,401 - synapse.server - 312 - INFO - main - Finished setting up.
2021-08-17 19:01:45,411 - synapse.push.pusher - 38 - INFO - main - email enable notifs: False
2021-08-17 19:01:45,417 - synapse.handlers.auth - 213 - INFO - main - Extra password_providers: []
2021-08-17 19:01:45,424 - synapse.federation.federation_server - 1075 - INFO - main - Registering federation EDU handler for 'm.device_list_update'
2021-08-17 19:01:45,426 - synapse.federation.federation_server - 1095 - INFO - main - Registering federation query handler for 'profile'
2021-08-17 19:01:45,433 - synapse.federation.federation_server - 1075 - INFO - main - Registering federation EDU handler for 'm.presence'
2021-08-17 19:01:45,434 - synapse.federation.federation_server - 1075 - INFO - main - Registering federation EDU handler for 'm.typing'
2021-08-17 19:01:45,441 - synapse.federation.federation_server - 1095 - INFO - main - Registering federation query handler for 'directory'
2021-08-17 19:01:45,454 - twisted - 258 - INFO - main - Redirected stdout/stderr to logs
2021-08-17 19:01:45,454 - synapse.app.homeserver - 118 - INFO - sentinel - Running
2021-08-17 19:01:45,457 - synapse.app.homeserver - 29 - INFO - sentinel - Set file limit to: 1048576
2021-08-17 19:01:45,458 - synapse.handlers.deactivate_account - 207 - INFO - user_parter_loop-0 - Starting user parter
2021-08-17 19:01:45,459 - synapse.config.tls - 308 - INFO - sentinel - Loading TLS key from /conf/server.tls.key
2021-08-17 19:01:45,460 - synapse.config.tls - 295 - INFO - sentinel - Loading TLS certificate from /conf/server.tls.crt
2021-08-17 19:01:45,481 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.health.HealthResource object at 0x7f70dad0f760> to path b'/health'
2021-08-17 19:01:45,481 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.federation.transport.server.TransportLayerServer object at 0x7f70dad0f820> to path b'/_matrix/federation'
2021-08-17 19:01:45,482 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.media.v1.media_repository.MediaRepositoryResource object at 0x7f70dad0f970> to path b'/_matrix/media/r0'
2021-08-17 19:01:45,482 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.media.v1.media_repository.MediaRepositoryResource object at 0x7f70dad0f970> to path b'/_matrix/media/v1'
2021-08-17 19:01:45,482 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.key.v2.KeyApiV2Resource object at 0x7f70da4b5490> to path b'/_matrix/key/v2'
2021-08-17 19:01:45,483 - twisted - 258 - INFO - sentinel - SynapseSite (TLS) starting on 8448
2021-08-17 19:01:45,484 - synapse.app.homeserver - 156 - INFO - sentinel - Synapse now listening on TCP port 8448 (TLS)
2021-08-17 19:01:45,514 - synapse.federation.federation_server - 1075 - INFO - sentinel - Registering federation EDU handler for 'm.receipt'
2021-08-17 19:01:45,516 - synapse.federation.federation_server - 1075 - INFO - sentinel - Registering federation EDU handler for 'org.matrix.signing_key_update'
2021-08-17 19:01:45,516 - synapse.federation.federation_server - 1095 - INFO - sentinel - Registering federation query handler for 'client_keys'
2021-08-17 19:01:45,517 - synapse.federation.federation_server - 1075 - INFO - sentinel - Registering federation EDU handler for 'm.direct_to_device'
2021-08-17 19:01:45,523 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.health.HealthResource object at 0x7f70da4b59a0> to path b'/health'
2021-08-17 19:01:45,523 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.ClientRestResource object at 0x7f70da4c01f0> to path b'/_matrix/client/api/v1'
2021-08-17 19:01:45,524 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.ClientRestResource object at 0x7f70da4c01f0> to path b'/_matrix/client/r0'
2021-08-17 19:01:45,524 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.ClientRestResource object at 0x7f70da4c01f0> to path b'/_matrix/client/unstable'
2021-08-17 19:01:45,524 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.ClientRestResource object at 0x7f70da4c01f0> to path b'/_matrix/client/v2_alpha'
2021-08-17 19:01:45,525 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.ClientRestResource object at 0x7f70da4c01f0> to path b'/_matrix/client/versions'
2021-08-17 19:01:45,525 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.well_known.WellKnownResource object at 0x7f70da47acd0> to path b'/.well-known/matrix/client'
2021-08-17 19:01:45,525 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.admin.AdminRestResource object at 0x7f70da47ad60> to path b'/_synapse/admin'
2021-08-17 19:01:45,526 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.synapse.client.pick_idp.PickIdpResource object at 0x7f70da48b160> to path b'/_synapse/client/pick_idp'
2021-08-17 19:01:45,526 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <twisted.web.resource.Resource object at 0x7f70da48b1c0> to path b'/_synapse/client/pick_username'
2021-08-17 19:01:45,526 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.synapse.client.new_user_consent.NewUserConsentResource object at 0x7f70da48b520> to path b'/_synapse/client/new_user_consent'
2021-08-17 19:01:45,527 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.synapse.client.sso_register.SsoRegisterResource object at 0x7f70da48b4f0> to path b'/_synapse/client/sso_register'
2021-08-17 19:01:45,527 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching FilePath('/usr/local/lib/python3.8/site-packages/synapse/static') to path b'/_matrix/static'
2021-08-17 19:01:45,527 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.media.v1.media_repository.MediaRepositoryResource object at 0x7f70dad0f970> to path b'/_matrix/media/r0'
2021-08-17 19:01:45,527 - synapse.util.httpresourcetree - 46 - INFO - sentinel - Attaching <synapse.rest.media.v1.media_repository.MediaRepositoryResource object at 0x7f70dad0f970> to path b'/_matrix/media/v1'
2021-08-17 19:01:45,528 - twisted - 258 - INFO - sentinel - SynapseSite starting on 8008
2021-08-17 19:01:45,528 - synapse.app.homeserver - 165 - INFO - sentinel - Synapse now listening on TCP port 8008
2021-08-17 19:01:45,593 - synapse.storage.background_updates - 109 - INFO - background_updates-0 - Starting background schema updates
2021-08-17 19:01:45,594 - synapse.handlers.deactivate_account - 217 - INFO - user_parter_loop-0 - User parter finished: stopping
2021-08-17 19:01:45,595 - synapse.push.pusherpool - 320 - INFO - start_pushers-0 - Started pushers
2021-08-17 19:01:45,601 - synapse.access.http.8008 - 389 - INFO - GET-0 - ::ffff:172.29.0.1 - 8008 - {None} Processed request: 0.000sec/-0.000sec (0.001sec, 0.000sec) (0.000sec/0.000sec/0) 394B 200 "GET /_matrix/client/versions HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:45,980 - synapse.access.http.8008 - 389 - INFO - POST-2 - ::ffff:172.29.0.1 - 8008 - {@alice:hs1} Processed request: 0.369sec/-0.000sec (0.019sec, 0.010sec) (0.185sec/0.032sec/31) 37B 200 "POST /_matrix/client/r0/createRoom HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:46,000 - synapse.http.servlet - 395 - WARNING - POST-3 - Unable to parse JSON: Expecting value: line 1 column 1 (char 0) (b'')
2021-08-17 19:01:46,098 - synapse.access.http.8008 - 389 - INFO - POST-1 - ::ffff:172.29.0.1 - 8008 - {@alice:hs1} Processed request: 0.488sec/-0.000sec (0.022sec, 0.005sec) (0.232sec/0.039sec/40) 77B 200 "POST /_matrix/client/r0/createRoom HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:46,154 - synapse.access.http.8008 - 389 - INFO - POST-3 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 0.170sec/-0.000sec (0.006sec, 0.001sec) (0.121sec/0.008sec/13) 37B 200 "POST /_matrix/client/r0/join/%21QFQwTSIvHJjnwOFTAu:hs1 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:46,234 - synapse.access.http.8008 - 389 - INFO - POST-4 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 0.132sec/-0.000sec (0.007sec, 0.001sec) (0.096sec/0.006sec/11) 37B 200 "POST /_matrix/client/r0/join/%23room_alias_random_2:hs1 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:46,238 - synapse.util.caches.response_cache - 205 - INFO - GET-6 - [sync]: using incomplete cached result for [(@bob:hs1, 1000, None, None, False, 'ICZLJCRTJT')]
2021-08-17 19:01:46,299 - synapse.access.http.8008 - 389 - INFO - GET-6 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 0.062sec/-0.000sec (0.004sec, 0.001sec) (0.000sec/0.000sec/0) 8176B 200 "GET /_matrix/client/r0/sync?timeout=1000 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:46,304 - synapse.access.http.8008 - 389 - INFO - GET-5 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 0.145sec/-0.000sec (0.019sec, 0.005sec) (0.121sec/0.019sec/32) 8176B 200 "GET /_matrix/client/r0/sync?timeout=1000 HTTP/1.1" "Go-http-client/1.1" [7 dbevts]
2021-08-17 19:01:46,312 - synapse.http.servlet - 395 - WARNING - POST-8 - Unable to parse JSON: Expecting value: line 1 column 1 (char 0) (b'')
2021-08-17 19:01:46,342 - synapse.access.http.8008 - 389 - INFO - GET-7 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 0.032sec/-0.000sec (0.006sec, 0.001sec) (0.008sec/0.002sec/11) 2898B 200 "GET /_matrix/client/r0/sync?since=s20_3_0_1_1_1_1_3_1&timeout=1000 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:46,351 - synapse.access.http.8008 - 389 - INFO - POST-8 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 0.039sec/-0.000sec (0.005sec, 0.001sec) (0.003sec/0.005sec/6) 2B 200 "POST /_matrix/client/r0/rooms/%21QFQwTSIvHJjnwOFTAu:hs1/leave HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:46,389 - synapse.access.http.8008 - 389 - INFO - GET-9 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 0.033sec/-0.000sec (0.011sec, 0.002sec) (0.024sec/0.002sec/8) 8420B 200 "GET /_matrix/client/r0/sync?timeout=1000 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:46,597 - synapse.storage.background_updates - 232 - INFO - background_updates-0 - Starting update batch on background update 'add_rooms_room_version_column'
2021-08-17 19:01:46,607 - synapse.storage.background_updates - 268 - INFO - background_updates-0 - Running background update 'add_rooms_room_version_column'. Processed 100 items in 8ms. (total_rate=0/ms, current_rate=0/ms, total_updated=0, batch_size=100)
2021-08-17 19:01:48,035 - synapse.storage.background_updates - 232 - INFO - background_updates-0 - Starting update batch on background update 'add_rooms_room_version_column'
2021-08-17 19:01:48,041 - synapse.access.http.8008 - 389 - INFO - GET-10 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 1.648sec/-0.000sec (0.002sec, 0.001sec) (0.001sec/0.000sec/1) 343B 200 "GET /_matrix/client/r0/sync?since=s22_3_0_1_1_1_1_3_1&timeout=1000 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:48,051 - synapse.storage.background_updates - 268 - INFO - background_updates-0 - Running background update 'add_rooms_room_version_column'. Processed 1250 items in 9ms. (total_rate=12.5/ms, current_rate=12.5/ms, total_updated=100, batch_size=1250)
2021-08-17 19:01:49,058 - synapse.access.http.8008 - 389 - INFO - GET-11 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 1.010sec/-0.000sec (0.004sec, 0.001sec) (0.001sec/0.000sec/1) 343B 200 "GET /_matrix/client/r0/sync?since=s22_3_0_1_1_1_1_3_1&timeout=1000 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:49,059 - synapse.storage.background_updates - 232 - INFO - background_updates-0 - Starting update batch on background update 'current_state_events_membership'
2021-08-17 19:01:49,074 - synapse.storage.background_updates - 268 - INFO - background_updates-0 - Running background update 'current_state_events_membership'. Processed 19 items in 12ms. (total_rate=0/ms, current_rate=0/ms, total_updated=0, batch_size=100)
2021-08-17 19:01:50,073 - synapse.access.http.8008 - 389 - INFO - GET-12 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 1.007sec/-0.000sec (0.003sec, 0.000sec) (0.000sec/0.000sec/1) 343B 200 "GET /_matrix/client/r0/sync?since=s22_3_0_1_1_1_1_3_1&timeout=1000 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:50,082 - synapse.storage.background_updates - 232 - INFO - background_updates-0 - Starting update batch on background update 'delete_old_current_state_events'
2021-08-17 19:01:50,084 - synapse.storage.databases.main.state - 409 - INFO - background_updates-0 - skipping rooms which are being created: set()
2021-08-17 19:01:50,084 - synapse.storage.databases.main.state - 425 - INFO - background_updates-0 - Deleting current state left rooms: set()
2021-08-17 19:01:50,091 - synapse.storage.background_updates - 268 - INFO - background_updates-0 - Running background update 'delete_old_current_state_events'. Processed 100 items in 8ms. (total_rate=0/ms, current_rate=0/ms, total_updated=0, batch_size=100)
2021-08-17 19:01:51,087 - synapse.access.http.8008 - 389 - INFO - GET-13 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 1.007sec/-0.000sec (0.004sec, 0.000sec) (0.000sec/0.000sec/1) 343B 200 "GET /_matrix/client/r0/sync?since=s22_3_0_1_1_1_1_3_1&timeout=1000 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:51,094 - synapse.storage.background_updates - 232 - INFO - background_updates-0 - Starting update batch on background update 'delete_old_current_state_events'
2021-08-17 19:01:51,102 - synapse.storage.background_updates - 268 - INFO - background_updates-0 - Running background update 'delete_old_current_state_events'. Processed 1250 items in 6ms. (total_rate=12.5/ms, current_rate=12.5/ms, total_updated=100, batch_size=1250)
2021-08-17 19:01:52,105 - synapse.access.http.8008 - 389 - INFO - GET-14 - ::ffff:172.29.0.1 - 8008 - {@bob:hs1} Processed request: 1.012sec/-0.000sec (0.005sec, 0.000sec) (0.001sec/0.000sec/1) 343B 200 "GET /_matrix/client/r0/sync?since=s22_3_0_1_1_1_1_3_1&timeout=1000 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-08-17 19:01:52,108 - synapse.storage.background_updates - 232 - INFO - background_updates-0 - Starting update batch on background update 'devices_last_seen'
2021/08/18 00:31:52 ============== 227d4b5cbe443c7cd67db01bcb6bc6db3c02b78a6f640d9a4f69bc84da188624 : END LOGS ==============
```

Reply:
I am unable to figure out what is the issue here as the similar thing is done in the sytests 😕

Reply:
The problem is `SyncUntilTimelineHas` only works for events when you're joined to a room. If you look at the [sync response](https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-r0-sync) you can see that:

```
leave | {string: Left Room} | The rooms that the user has left or been banned from, mapped as room ID to room information.
```

..but [SyncUntilTimelineHas](https://github.com/matrix-org/complement/blob/master/internal/client/client.go#L135) only looks in the `join` section. Instead, you need to use `SyncUntil`to look into `"rooms.leave."+GjsonEscape(roomID)+".timeline.events"`

Reply:
@kegsay still getting the same logs after using `"rooms.leave."+GjsonEscape(roomID)+".timeline.events"`
```sh
    apidoc_room_members_test.go:214: Making GET request to http://localhost:55077/_matrix/client/r0/sync
=== CONT  TestRoomMembers
    client.go:370: GET hs1/_matrix/client/r0/sync => 200 OK (1.014066563s)
=== CONT  TestRoomMembers/Parallel/POST_/rooms/:room_id/leave_can_leave_a_room
    apidoc_room_members_test.go:214: HTTP/1.1 200 OK
        Transfer-Encoding: chunked
        Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, Date
        Access-Control-Allow-Methods: GET, HEAD, POST, PUT, DELETE, OPTIONS
        Access-Control-Allow-Origin: *
        Cache-Control: no-cache, no-store, must-revalidate
        Content-Type: application/json
        Date: Fri, 20 Aug 2021 19:57:20 GMT
        Server: Synapse/1.37.1

        157
        {"account_data":{"events":[]},"to_device":{"events":[]},"device_lists":{"changed":[],"left":[]},"presence":{"events":[]},"rooms":{"join":{},"invite":{},"knock":{},"leave":{}},"groups":{"join":{},"invite":{},"leave":{}},"device_one_time_keys_count":{},"org.matrix.msc2732.device_unused_fallback_key_types":[],"next_batch":"s15_3_0_1_1_1_1_3_1"}
        0

```